lockfile/src/lib.rs

```diff
@@ -0,0 +1,247 @@
+        self.serialize(f)
+    }
+}
+
+impl LockfileFormat {
+    /// Get the canonical Phylum name for this format.
+    ///
+    /// This is the string used in documentation and examples where the user
+    /// specifies a lock file format by name.
+    ///
+    /// This method returns the same value as `.to_string()`, but is const and
+    /// returns a `&'static str`.
+    pub const fn name(&self) -> &'static str {
+        match self {
+            LockfileFormat::Yarn => "yarn",
+            LockfileFormat::Npm => "npm",
+            LockfileFormat::Gem => "gem",
+            LockfileFormat::Pip => "pip",
+            LockfileFormat::Pipenv => "pipenv",
+            LockfileFormat::Poetry => "poetry",
+            LockfileFormat::Maven => "mvn",
+            LockfileFormat::Gradle => "gradle",
+            LockfileFormat::Msbuild => "nuget",
+        }
+    }
+
+    /// Get the corresponding parser for the specified format.
+    pub fn parser(&self) -> &'static dyn Parse {
+        match self {
+            LockfileFormat::Yarn => &YarnLock,
+            LockfileFormat::Npm => &PackageLock,
+            LockfileFormat::Gem => &GemLock,
+            LockfileFormat::Pip => &PyRequirements,
+            LockfileFormat::Pipenv => &PipFile,
+            LockfileFormat::Poetry => &Poetry,
+            LockfileFormat::Maven => &Pom,
+            LockfileFormat::Gradle => &GradleLock,
+            LockfileFormat::Msbuild => &CSProj,
+        }
+    }
+
+    /// Iterate over all supported lock file formats.
+    pub fn iter() -> LockfileFormatIter {
+        LockfileFormatIter(0)
+    }
+}
+
+/// An iterator of all supported lock file formats.
+pub struct LockfileFormatIter(u8);
+
+impl Iterator for LockfileFormatIter {
+    type Item = LockfileFormat;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let item = match self.0 {
+            0 => LockfileFormat::Yarn,
+            1 => LockfileFormat::Npm,
+            2 => LockfileFormat::Gem,
+            3 => LockfileFormat::Pip,
+            4 => LockfileFormat::Pipenv,
+            5 => LockfileFormat::Poetry,
+            6 => LockfileFormat::Maven,
+            7 => LockfileFormat::Gradle,
+            8 => LockfileFormat::Msbuild,
+            _ => return None,
+        };
+        self.0 += 1;
+        Some(item)
+    }
+}
+
+pub type ParseResult = anyhow::Result<Vec<PackageDescriptor>>;
+
+pub trait Parse {
+    /// Parse from a string
+    fn parse(&self, data: &str) -> ParseResult;
+
+    /// Indicate the type of file parsed by this parser
+    fn format(&self) -> LockfileFormat;
+
+    /// Indicate the type of packages parsed by this parser
+    fn package_type(&self) -> PackageType;
+
+    /// Test if a file name could be a lock file supported by this parser.
+    ///
+    /// The file does not need to exist.
+    fn is_path_lockfile(&self, path: &Path) -> bool;
+}
+
+/// Get the parser of a potential lock file.
```

Comment:
```suggestion
/// Get the parser for a potential lock file.
```

Your english is most likely much better than mine. But I think this reads a little nicer?

Reply:
`for` makes more sense if it's getting a parser. Maybe at some point it was getting the expected format instead.